### ISSUE TITLE
Asynchronously rendered prompt?

### ISSUE THREAD

**Issue by iirelu** [fish-shell/fish-shell#1942]
It's bugged me for a while; *Nothing* is rendered until *both* `fish_prompt` and `fish_right_prompt` have finished, which means that for [prompts that take a while](https://github.com/milkbikis/powerline-shell) there is an extraordinarily noticeable latency between pressing enter and being able to type. Would it really be impossible to "stream" the prompt in and reflow any already-written commands whenever its length changes?

A [few other shells](http://www.anishathalye.com/2015/02/07/an-asynchronous-shell-prompt/) allow asynchronous prompt rendering to an extent; they don't wait on the right prompt to finish rendering. This would definitely be an improvement over what we have now.

**Comment by ridiculousfish** [comment]
Eventually, but first we need the capability to run fish code in the background!

**Comment by d12frosted** [comment]
Firstly, thank you very much for that link. Willing to play with his configs. Secondly, I find this idea very good. Sometimes (for whatever reason) I am stuck while fish is trying to render my prompt. Yeah, that's my bad of making it hard-to-evaluate. But anyway.

**Comment by mafredri** [comment]
This is a simple workaround for performing async tasks in the prompt that I started sketching up today after trying out fish for the first time. Feel free to extend it if you like:

##### prompt.fish
```fish
function --on-signal WINCH handler
    set -l mypid %self
    set -l value (eval echo "\$prompt_$mypid")
    set -g _fish_prompt_updated (echo " (updated $value)")
end

function fish_prompt
    ....
    set -l updated "$_fish_prompt_updated"
    if test "$updated" = ""
        ./async.fish %self &
    end
    set -g _fish_prompt_updated ""

    echo -n $updated
    ...
end
```

##### async.fish
```
#!/usr/bin/env fish
set caller $argv[1]

sleep 1
set -U prompt_$caller (date)
kill -WINCH $caller
```

**Comment by psychoticmeow** [comment]
An interesting idea @mafredri, I took it and simplified it a little since it doesn't appear necessary to use the handler function:

**prompt.fish**
```fish
function fish_prompt
	set -l caller %self
	set -l user (whoami)
	set -l host (hostname|cut -d . -f 1)
	set -l dir (prompt_pwd)
	set -l updated (eval echo "\$prompt_$caller")
	set -U prompt_{$caller} ""

	if test $updated = ""
		/path/to/async.fish %self &
	end

	printf "%s at %s in %s%s" \
		(set_color yellow)(echo $user)(set_color normal) \
		(set_color blue)(echo $host)(set_color normal) \
		(set_color $fish_color_cwd)(echo $dir)(set_color normal) \
		(echo $updated)
	echo "↪ "
end
```

**async.fish**
```fish
#!/usr/bin/env fish
set caller $argv[1]

function append --description 'Append text to the end of the prompt.'
	set -U prompt_{$caller} (eval echo "\$prompt_$caller")(echo "$argv")
end

function refresh --description 'Refresh the prompt to show changes.'
	kill -WINCH $caller
end

sleep 1
append ' '(date)
refresh
```

However with version 2.1.2 I'll eventually get a segfault, or if I hold down enter I'll very quickly get a segfault, it doesn't matter who's script I'm using.

Some other caveats:
* Need to either `cut` or `fold` the result because if the line gets too long the whole thing just gets truncated to a '>' prompt.
* You need to wait for the async part to finish before using history search, because every time the prompt is changed whatever text is after the prompt will be considered something you have typed and not an item from your history.

**Comment by mafredri** [comment]
Nice, glad you found some use for it @psychoticmeow!

Regarding the segfault, that's an interesting find. I can actually reproduce it (fish v2.1.1, osx) using your code samples (as is) but not using mine (as is). But the shell segfaulting is definitely a bug, that should not happen from simply running scripts.

**Comment by psychoticmeow** [comment]
@mafredri I've updated my scripts, I'm trying to allow for slower commands to be run last, without holding up the display of other useful information like the current git branch. However I can't quite get it working segfault free:

https://gist.github.com/psychoticmeow/d5d432aab719cc073ae9

When line 13 is uncommented the partial updates work, but it will definitely segfault at some point. I guess that the async script is updating it at the same time that the prompt script is reading it.

**Comment by stefanmaric** [comment]
Hi guys, any news on this topic? It seems a bit stale and wanted to know if there are improvements for this and/or #23.

**Comment by ridiculousfish** [comment]
Unfortunately this is a very big effort that is probably blocked by #563. No work has been done yet.

**Comment by acomagu** [comment]
Hi, I created plugin make prompt asynchronous easily.
[acomagu/fish-async-prompt](https://github.com/acomagu/fish-async-prompt)
I'm happy if you could try it and give comment!

**Comment by alphapapa** [comment]
@acomagu That looks interesting.  Could you explain how it basically works?

**Comment by acomagu** [comment]
@alphapapa Almost same logic as https://github.com/fish-shell/fish-shell/issues/1942#issuecomment-84626212.
It forks new fish process to calculate prompt string on every <Enter> key(handling with `fish_prompt` event), after the end, the child process refreshes the parent's prompt with WINCH signal.
Fish refreshes prompt when receiving WINCH signal.(undocumented behavior)

**Comment by acomagu** [comment]
@alphapapa And it replaces `fish_right_prompt` and `fish_prompt` functions with own on startup.

**Comment by alphapapa** [comment]
@acomagu  Thanks.  Does it suffer from any segfault issues as those comments described?  i.e. if I `cd` into a large git repo and then start pressing `RET` before the git prompt has had time to return and render, does it segfault, or does it build up a list of unfinished prompt-rendering processes?

**Comment by acomagu** [comment]
@alphapapa I have not met segfault issue on developing my plugin, or with the code on the comment.

Now implementation will spawn many processes if we holding `<Return>` key.
I'm not feeling stress for it so leave it for now, but it can be avoided by adding one another universal variable, like `fish_prompt_update_running` of https://github.com/fish-shell/fish-shell/issues/1942#issuecomment-84473886. (So if someone needs it, I'll add it)

I think the most big problem the plugin has currently is, when execute `exit` command and if the spawned process is living yet, it prints warning (`There are still jobs active (use the jobs command to see them). A second attempt to exit will terminate them.`)
I don't have idea to solve this smartly yet.

**Comment by alphapapa** [comment]
Thanks.  I think it would be important to solve those before expecting it to be usable by most users.

I guess you might be able to solve the second one by trapping on EXIT (or using Fish's on-event equivalent) and killing any prompt-making processes before Fish has a chance to complain.

**Comment by acomagu** [comment]
@alphapapa That make sense. I'll try it later.
Thank you for good idea!

**Comment by c02y** [comment]
Hi, https://github.com/acomagu/fish-async-prompt is nice, but it will always break when fish releases a new version. So I think it is better that the developers of fish shell implement this feature inside fish shell project.

**Comment by ridiculousfish** [comment]
Yeah async prompts and completions will be a big client of async execution when it lands.

**Comment by NonLogicalDev** [comment]
In huge monorepos it is quite important to have a way to render out prompt asynchronously. I feel like ZSH handled this very elegantly by allowing a prompt process to hook up a file descriptor and a corresponding handler for the file descriptor to it's interactive line editor (see `zle -F` in https://zsh.sourceforge.io/Doc/Release/Zsh-Line-Editor.html). Doing so causes ZLE to re-exec the handler on every poll event from file descriptor which has full access to current shell session and can modify the prompt variables. Overall super clean, as this does not involve firing off background jobs.

I don't think that is a very well known feature in ZSH, but it enabled me to create a super responsive prompt for it: https://github.com/NonLogicalDev/shell.async-goprompt

I am really curious if I could implement something similar to this for `fish`. I think @acomagu's firing a WINCH signal idea is promising, but certainly a bit less elegant than a file descriptor.

**Comment by acomagu** [comment]
Just note: Several years after its initial implementation, the implementation approach of fish-async-prompt has changed a bit. Instead of sending `WINCH`, now it repaints the shell by executing `commandline -f repaint` when the prompt string is ready.

**Comment by ilan-schemoul** [comment]
The solution here is hard to understand for me https://github.com/acomagu/fish-async-prompt. I really need to apply to my own prompt but the code is unclear (I read the original comment a bit but not sure yet how to apply it simply to my code).

Anyway I had an idea. But I would need a way to know the command number (e.g: it's the third command I type since the start of fish I can get echo $fish_command_number => 3).

Once i know how to do that  it's trivial to first echo a basic prompt then call call_to_git (with expensive git calls) and then redraw the prompt (but this call_to_git knows to not compute again because he knows he already computed $fish_command_number)
```fish
call_to_git
   if test last_command_processed = (echo $fish_command_number)
      return
   end
  last_command_processed = (echo $fish_command_number)
  set git_infos (git status)
  # calls fish_prompt
  commandline -f repaint

fish_prompt
  # First git_infos is empty string. Second time it will have actual data.
  echo $git_infos
  echo $prompt
  # First git_infos is called. Second it's called but will skip another round of computations.
  call_to_git
```

Do someone know how to get something similar to my idea (fish_command_number) ?

**Comment by ilan-schemoul** [comment]
Actually even simpler: can we have a way to determine fish_prompt is called because of "redraw". Because then things are even easier !

**Comment by NonLogicalDev** [comment]
I managed to get the async rendering accomplished via universal variables in Fish combined with `--on-variable` hooks.

I found that to be a fairly nice and simple way to do that, maybe just a smidge slower than ZSH's file descriptor integration, but it does not hurt the responsiveness.

Honestly I feel like this is a way cleaner, more readable approach.

If anyone is interested I have the implementation here for my prompt:
https://github.com/NonLogicalDev/shell.async-goprompt/blob/main/plugin/fish/prompt_async_setup.fish
